File: fibonacci/fibonacci.py
class Solution:
    memo_list = [0] * 31
    def fib(self, n: int) -> int:
        if (n == 0 or n == 1):
            return n
        if (self.memo_list[n] is not 0):
            return self.memo_list[n]
        self.memo_list[n] = self.fib(n - 2) + self.fib(n - 1)
        return self.memo_list[n]

File: fibonacci/fibonacci.java
/**
 * returns the fibonacci nth number
 * this implementation uses a topdown approach
 * base case: n = 0 return 0, n = 1 returns 1
 * To avoid redundant calculations, answers are cached in arr, where arr[n] gives the nth
 * fibonacci number
 * */
class fibonacci {
    int arr[] = new int[31];
    public int fib(int n) {
        if (n == 0) {
            return 0;
        }
        if (n == 1) {
            return 1;
        }

        if (arr[n] != 0) {
            return arr[n];
        }
        arr[n] = fib(n - 2) + fib(n - 1);
        return arr[n];
    }
}
